### PR TITLE
Represent features as a BTreeMap

### PR DESCRIPTION
**Pull request by danieldk**
This removes the dual String/BTreeMap representation of features,
where the map was lazily initialized.

The dual representation setup was clunky and made it impossible to
change the feature map in-place. This representation was initially
chosen to avoid the feature parsing overhead. However, the difference
in performance is fairly small:

Old: 4.446
New: 4.776

@sebpuetz I'll merge this if you do not have strong objections.

### CODE REVIEW

File: src/token.rs [hunk 1]
```diff
@@ -278,35 +238,30 @@ where
...
 
-fn map_to_string(feature_map: &BTreeMap<String, Option<String>>) -> String {
-    feature_map
-        .iter()
-        .map(|(k, v)| match *v {
-            Some(ref v) => format!("{}:{}", k, v),
-            None => k.to_owned(),
-        })
-        .join("|")
+impl From<&Features> for String {
```
Comment (sebpuetz) [inline src/token.rs:254]: Wouldn't `ToString` be more expected here?

**Review by sebpuetz — COMMENTED** [review]
Guess I'm late to the party, but anyways, I think `ToString` could also be a candidate to get `String` from `&Features`.

File: src/token.rs (hunk 1 above)
Comment (danieldk) [reply src/token.rs:254]: `ToString` is already implemented through `Display`. Another question is why we need this if we have `Display` ;).

Comment (sebpuetz) [reply src/token.rs:254]: Well, then... 

I was only looking for implementors of `ToString` expecting to find something like `impl<T> ToString for T where String: From<&T>`, didn't even bother to check for a `Display` impl on `Token` ;)

**Pull request merged**